### PR TITLE
Implement grouping functionality 

### PR DESCRIPTION
**Pull request by CMDRyan**
@Steven-Chan 

UI changes
- Profile icon has been replaced with hamburger icon
- Hamburger icon will trigger curtain menu which includes groupby dropdown box

Added features
- Library can now handle groupBy function call and will return data to client in an array format
- GroupBy function has been added
   - By machine type
   - By Project name

To-Do:
- Align number designs

### CODE REVIEW

File: src/getData.js [hunk 1]
```diff
@@ -12,16 +12,25 @@ module.exports.getProjectIDs = function getProjectIDs(requestedGroupID, token) {
...
-      const ids = projects.map((project) => project.id);
-
-      return ids;
+      const projectList = [];
+      projects.map((project) => {
+        Project = {
+          id: project.id,
+          name: project.name,
+        };
+        projectList.push(Project);
```
Comment (Steven-Chan) [inline src/getData.js:21]: Bad code style. Missing "const". Variable name should be camelcase starts with small letters.

Comment (Steven-Chan) [reply src/getData.js:21]: I just found this style also appears somewhere else, pls help fix them altogether.

**Review by Steven-Chan — CHANGES_REQUESTED** [review]

**Comment by CMDRyan** [comment]
@Steven-Chan 

- Added const to object decorations
- Refactored groupBy "none", following the same style as other criteria
- Refactored group titles html building method using `Template` element
- Modified title css to compatible with the `Template` element

**Comment by CMDRyan** [comment]
@Steven-Chan

Changes to UI:

- Changed feedback for cards to colour changing (white to light gray)
- Removed curtain navigation menu

Screenshot 2021-09-05 at 9 59 23 PM

Screenshot 2021-09-05 at 10 01 01 PM

File: src/views/partials/jobCard.ejs [hunk 2]
```diff
@@ -1,17 +1,35 @@
...
-    Started At: <span class="text-gray-400 relative">
-      <%= job[current].started_at %>
-    </span><br>
-    tags: <span class="text-gray-400 relative">
-      <%= job[current].tag_list %>
-    </span>
+  <div class="<%=job[current].status%>-card flex flex-col mt-0.5 mx-0.5 sm:p-5 sm:h-26 p-1 min-h-0 w-auto border-solid border-1 text-gray-400 bg-gray-50 transition duration-100 hover:bg-gray-400">
+
+    <div class="flex flex-row">
+      <div class="flex flex-grow-0">Project:</div>
```
[the remark this answers is not in the document]
Comment (Steven-Chan) [reply src/views/partials/jobCard.ejs:5]: Also you only specify the flex-grow, but not flex-shrink. Most cases I think we don't specify flex-grow or flex-shrink separately but use flex-1 directly. If the element has a fixed size, i think you better want to give it a fix width instead.

File: src/views/partials/singleJobCard.ejs [hunk 3]
```diff
@@ -1,17 +1,35 @@
...
-      Created At: <span class="text-gray-400 relative">
-        <%= job.created_at %>
-      </span><br>
-      Started At: <span class="text-gray-400 relative">
-        <%= job.started_at %>
-      </span><br>
-      tags: <span class="text-gray-400 relative">
-        <%= job.tag_list %>
-      </span>
+  <a href="<%=job.web_url%>" target="_blank">
```
Comment (Steven-Chan) [inline src/views/partials/singleJobCard.ejs:1]: I just realize there is a `jobCard.js` and `singleJobCard.js`. Can you help refactor and make them one single template file.

File: src/views/pages/index.ejs [hunk 4]
```diff
@@ -63,5 +94,73 @@
...
+    }
+  </script>
+  <template>
+    <div class="flex flex-row">
+      <div class="flex flex-grow"></div>
+      <div class="flex flex-grow-0"></div>
+      <div class="flex flex-grow-0"></div>
+      <div class="flex flex-grow"></div>
+    </div>
+  </template>
```
Comment (Steven-Chan) [inline src/views/pages/index.ejs:164]: What is this?

File: src/views/pages/index.ejs [hunk 5]
```diff
@@ -63,5 +94,73 @@
...
+      var htmlArr = []
+      groups.forEach(groupName => {
+        const groupJobs = (data.filter(job=>job[criteria] == groupName))
+        var titleTemplate = $("Template")[0]
+        var titleItem = titleTemplate.content.cloneNode(true)
+        var titleDiv = titleItem.querySelectorAll("div")
+        titleDiv[0].classList.add(`${this.id}-group-title`)
+        titleDiv[2].innerHTML = groupName
+        titleDiv[3].classList.add(`${this.id}-circle-sm`)
+        titleDiv[3].innerHTML = groupJobs.length
```
Comment (Steven-Chan) [inline src/views/pages/index.ejs:150]: Don't use index to access the element.

**Review by Steven-Chan — CHANGES_REQUESTED** [review]

**Comment by Steven-Chan** [comment]
@CMDRyan  Pls also include screenshot with card.

**Comment by CMDRyan** [comment]
Full screen: (hovering on one card):
Screenshot 2021-09-08 at 11 48 16 AM

Mobile size (grouped by):
Screenshot 2021-09-08 at 11 48 59 AM

**Review by Steven-Chan — CHANGES_REQUESTED** [review]
@CMDRyan 
- The section header hight needs larger, it is too tight now
- The job count in section header is not aligned
- The job card spacing looks different in narrow screen and wide screen

**Comment by CMDRyan** [comment]
@Steven-Chan 

- The cards rendering are refactored, and the extra jobCards.js is removed
- The header section padding has been increased
- Job card spacing in mobile view has been fixed
- Template function no longer uses index method to change element

**Comment by CMDRyan** [comment]
Margin fixed for mobile view: 

Screenshot 2021-09-09 at 11 33 21 AM

**Comment by CMDRyan** [comment]
Grouping header paddings and number fix:
Screenshot 2021-09-09 at 1 03 18 PM

File: src/app.js [hunk 6]
```diff
@@ -108,10 +108,46 @@ app.get("/groups/:id/jobs", function (req, res) {
...
+        }
+      pendingCards.push(jobObj)
+      })
+      runningJobs.map((job)=>{
+        const jobObj = {
+          name: job.project_name,
+          tags: job.tag_list,
+          html: cardTemplate({job: job})
+        }
+      runningCards.push(jobObj)
```
Comment (Steven-Chan) [inline src/app.js:142]: This is not the correct way to use map. You transform the value of an array with map and the output of map is the array of new value already.

File: src/app.js [hunk 7]
```diff
@@ -108,10 +108,46 @@ app.get("/groups/:id/jobs", function (req, res) {
...
     .then((data) => {
+      const createdCards = []
+      const pendingCards = []
+      const runningCards = []
+      const createdJobs = data.filter((data) => data.status === "created")
+      const pendingJobs = data.filter((data) => data.status === "pending")
+      const runningJobs = data.filter((data) => data.status === "running")
+      const cardTemplate = ejs.compile(
+        read("src/views/partials/singleJobCard.ejs", "utf-8")
+      );
```
Comment (Steven-Chan) [inline src/app.js:119]: The template does not change in every request, so we should load it globally once.

**Review by Steven-Chan — CHANGES_REQUESTED** [review]

**Comment by CMDRyan** [comment]
@Steven-Chan 
- Card template has been made global
- Corrected logic of array mappings

**Pull request merged**